Jenkinsfile-utils.groovy

```diff
@@ -0,0 +1,356 @@
+#!/usr/bin/env groovy
+
+// General structure of this file:
+//
+// SECTION: Stage functions.
+// - stagePre()     : Function that should be invoked at the start of every stageFoo() function.
+// - stagePost()    : Function that should be invoked at the end of every stageFoo() function.
+// - stageFoo()     : A Jenkins stage.
+//
+// SECTION: Utility functions.
+// Random helper functions.
+//
+// You should probably know about Groovy's elvis operator ?:,
+// where a ?: b means
+//      if (a) { return a; } else { return b; }
+
+// SECTION: Stage functions.
+
+/** This should be invoked before every stage. */
+void stagePre() {
+    sh script: 'echo $NODE_NAME', label: 'Print node name.'
+    sh script: './build-support/print_docker_info.sh', label: 'Print image information.'
+}
+
+/** This should be invoked after every stage. */
+void stagePost() {
+    // No-op.
+}
+
+/** Test if the GitHub "ready-for-ci" label is present. Otherwise, abort the build. */
+void stageGithub() {
+    stagePre()
+    ready_for_build = sh script: 'python3 ./build-support/check_github_labels.py', returnStatus: true, label: 'Test Github labels.'
+    if (0 != ready_for_build) {
+        currentBuild.result = 'ABORTED'
+        error('Not ready for CI. Please add ready-for-ci tag in Github when you are ready to build your PR.')
+    }
+    stagePost()
+}
+
+/** Test if the codebase passes basic checks: format, documentation, lint, clang-tidy. */
+void stageCheck() {
+    stagePre()
+    installPackages('build')
+    sh 'cd apidoc && doxygen -u Doxyfile.in && doxygen Doxyfile.in 2>warnings.txt && if [ -s warnings.txt ]; then cat warnings.txt; false; fi'
+    sh 'mkdir build'
+    sh 'cd build && cmake -GNinja ..'
+    sh 'cd build && timeout 20m ninja check-format'
+    sh 'cd build && timeout 20m ninja check-lint'
+    sh 'cd build && timeout 20m ninja check-censored'
+    sh 'cd build && ninja check-clang-tidy'
+    stagePost()
+}
+
+/** Build and run the default "ninja" target.  */
+void stageBuildDefault(Map args = [:]) {
+    stagePre()
+    installPackages()
+    buildNoisePage(args)
+    stagePost()
+}
+
+/** Run the C++ unit tests, TPL tests, optionally generate coverage. */
+void stageTest(Boolean runPipelineMetrics, Map args = [:]) {
```

Comment:
The boolean is a little ugly, but this is because we don't uniformly want to run pipeline metrics on all stages. I'm actually not sure if we have a principled way of picking when to run pipeline metrics.

Reply:
Added comment to main Jenkinsfile.